tokio/src/runtime/task/list.rs

```diff
@@ -0,0 +1,33 @@
+use crate::loom::sync::Mutex;
+use crate::runtime::task::Task;
+use crate::util::linked_list::{Link, LinkedList};
+
+pub(crate) struct OwnedTasks<S: 'static> {
+    list: Mutex<LinkedList<Task<S>, <Task<S> as Link>::Target>>,
+}
+
+impl<S: 'static> OwnedTasks<S> {
+    pub(crate) fn new() -> Self {
+        Self {
+            list: Mutex::new(LinkedList::new()),
+        }
+    }
+
+    pub(crate) fn push_front(&self, task: Task<S>) {
+        self.list.lock().push_front(task);
+    }
+
+    pub(crate) fn pop_back(&self) -> Option<Task<S>> {
+        self.list.lock().pop_back()
+    }
+
+    /// The caller must ensure that if the provided task is stored in a
+    /// linked list, then it is in this linked list.
+    pub(crate) unsafe fn remove(&self, task: &Task<S>) -> Option<Task<S>> {
+        self.list.lock().remove(task.header().into())
+    }
```

Comment:
We had discussed introducing a field in the task header to remember which `OwnedTasks` it is in to make these operations safe. However this seems to be somewhat difficult, e.g. if you concurrently insert it into two `OwnedTasks` structures, then you have a race condition on the field for remembering the container.

Reply:
Right, which would be solved if the only way to get a `Task` structure is via the `OwnedTasks` value:

```rust
owned_tasks.insert(async { ... }) -> Task<_>;
```

This can be done in a follow up API though, we can keep `remove` unsafe for now.